### PR TITLE
Solution For "Regarding ability to override stripe api url per API re…

### PR DESCRIPTION
**Pull request by yasirshabbir44**
…quest #1181".

https://github.com/stripe/stripe-java/issues/1181

### CODE REVIEW

**Comment by richardm-stripe** [comment]
Hello @yasirshabbir44. Thank you for this change. It makes sense to me!

Can you sign the CLA per CLAassistant's comment above?

**Comment by yasirshabbir44** [comment]
Hello @richardm-stripe 
Done

**Comment by richardm-stripe** [comment]
Ah, I see you have changed `src/main/java/com/stripe/model/PaymentMethod.java` to replace `Stripe.getApiBase()` with `options.getApiBase()` but

- This file is generated, so we have to change the generator (which alas, is private) otherwise the change to this file will be reverted.
- Many other models also need to be changed to use `options.getApiBase()`.
- We need to account for the possibility that `options` is null, since passing `null` RequestOptions is allowed into methods like `PaymentMethod.create`.

So we can't merge as-is. I'll see if I can get a generator change in this week.

**Comment by yasirshabbir44** [comment]
Waiting for to get merge if you all get well please merge it.

thanks

**Comment by richardm-stripe** [comment]
Hello @yasirshabbir44, I'm working on this and made a number of changes:

* I changed `apiBase` to `baseUrl`. I wanted to give it a different name from `Stripe.apiBase` because `Stripe.connectBase` (which defaults to `connect.stripe.com`) and `Stripe.uploadBase` (which defaulst to `files.stripe.com`) also exist, but specifying `RequestOptions.baseUrl` will override *all* of these, not just `Stripe.apiBase`.
* I changed it so that defaulting to Stripe.getApiBase() happens in a new method `ApiResource.fullUrl` which is called from each resource method. It needs to be this way, since we need to handle null `RequestOptions`, and methods that use `connectBase` or `uploadBase` need a way to specify that they have different defaults from other methods.
* I changed our generator so that all methods in the SDK respect `RequestOptions.baseUrl` and committed the resulting changes.

File: src/main/java/com/stripe/net/ApiResource.java [hunk 1]
```diff
@@ -55,6 +55,14 @@ public ReflectionAccessFilter.FilterResult check(Class<?> rawClass) {
     return builder.create();
   }
 
+  public static String fullUrl(String defaultBaseUrl, RequestOptions options, String relativeUrl) {
```
Comment (pakrym-stripe) [inline src/main/java/com/stripe/net/ApiResource.java:58]:
```suggestion
  protected static String fullUrl(String defaultBaseUrl, RequestOptions options, String relativeUrl) {
```

Comment (richardm-stripe) [reply src/main/java/com/stripe/net/ApiResource.java:58]: Alas, StripeCollection is not a subclass of ApiResource

**Review by pakrym-stripe — APPROVED** [review]

**Comment by yasirshabbir44** [comment]
Could you please merge it

**Comment by richardm-stripe** [comment]
Sure, we can merge! We won't cut a release until the new year, though.

**Comment by richardm-stripe** [comment]
Released in https://github.com/stripe/stripe-java/releases/tag/v22.5.0

**Comment by yasirshabbir44** [comment]
Hello Richard,

I did changes and raised PR and that PR also got merged, but in the
contributor list I cannot see my name.

Why it’s not showing up?


Thanks
Yasir Shabbir

**Pull request merged**